server/functions/formatting.go

```diff
@@ -0,0 +1,1411 @@
+var errConflictingValues = errors2.NewKind(`conflicting values for "%s" field in formatting string`)
+
+// Time and date constants for formatting calculations
+const (
+	monthsPerYear = 12
+	hoursPerDay   = 24
+	minsPerHour   = 60
+	usecsPerSecs  = 1000000
+	maxTzdispHour = 15 // Maximum allowed hour part for timezone display
+)
+
+// formatNode represents a single element in a parsed format string
+type formatNode struct {
+	typ        formatNodeType // Type of the node (action, character, separator, space)
+	characters []uint8        // Characters for literal text nodes
+	suffix     *keywordSuffix // Prefix modifiers like FM, TM
+	postfix    *keywordSuffix // Suffix modifiers like TH, SP
+	key        *keyword       // Format keyword for action nodes
+}
+
+// formatNodeType defines the type of formatting node
+type formatNodeType uint
+
+const (
+	nodeTypeACTION    formatNodeType = iota + 1 // Format action (e.g., YYYY, MM, DD)
+	nodeTypeCHAR                                // Literal character
+	nodeTypeSEPARATOR                           // Separator character
+	nodeTypeSPACE                               // Whitespace
+)
+
+// fromCharDateMode defines the date parsing mode
+type fromCharDateMode uint
+
+const (
+	fromCharDateNONE      fromCharDateMode = iota // Value does not affect date mode
+	fromCharDateGREGORIAN                         // Gregorian (day, month, year) style date
+	fromCharDateISOWEEK                           // ISO 8601 week date
+)
+
+// tmFromChar holds parsed date/time components from input string
+// For char->date/time conversion
+type tmFromChar struct {
+	mode      fromCharDateMode // Date parsing mode (Gregorian, ISO week or neither)
+	hh        int              // Hour (12 or 24 hour format)
+	pm        int              // 0 if AM; 1 if PM
+	mi        int              // Minutes
+	ss        int              // Seconds
+	ssss      int              // Seconds since midnight
+	d         int              // Day of week (1-7, Sunday = 1, 0 means missing)
+	dd        int              // Day of month
+	ddd       int              // Day of year
+	mm        int              // Month
+	ms        int              // Milliseconds
+	year      int              // Year
+	bc        int              // BC/AD indicator
+	ww        int              // Week of year
+	w         int              // Week of month
+	cc        int              // Century
+	j         int              // Julian day
+	us        int              // Microseconds
+	yysz      int              // Year size (YY=2, YYYY=4)
+	is12clock bool             // True for 12-hour clock format
+	tzsign    int              // Timezone sign: +1, -1, or 0 if no TZH/TZM fields
+	tzh       int              // Timezone hours
+	tzm       int              // Timezone minutes
+	ff        int              // Fractional precision
+	//has_tz    bool             // True if TZ field was present
+	//gmtoffset int              // GMT offset of fixed-offset zone abbreviation
```

Comment:
Commented out? Were these supposed to be removed, or should we add a TODO for these?